Sparse CSR: enable autograd for torch.sparse.addmm and torch.sparse.mm

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #77243
* __->__ #76591

This PR updates the derivative rule for `torch.sparse.addmm` to be
working with CSR sparse matrix. Notably `torch.sparse.sampled_addmm` is
used in the backward function.

IvanYashchuk committed:
Sparse CSR: enable autograd for torch.sparse.addmm and torch.sparse.mm

This PR updates the derivative rule for `torch.sparse.addmm` to be
working with CSR sparse matrix. Notably `torch.sparse.sampled_addmm` is
used in the backward function.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: enable autograd for torch.sparse.addmm and torch.sparse.mm"


This PR updates the derivative rule for `torch.sparse.addmm` to be
working with CSR sparse matrix. Notably `torch.sparse.sampled_addmm` is
used in the backward function.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: enable autograd for torch.sparse.addmm and torch.sparse.mm"


This PR updates the derivative rule for `torch.sparse.addmm` to be
working with CSR sparse matrix. Notably `torch.sparse.sampled_addmm` is
used in the backward function.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: enable autograd for torch.sparse.addmm and torch.sparse.mm"


This PR updates the derivative rule for `torch.sparse.addmm` to be
working with CSR sparse matrix. Notably `torch.sparse.sampled_addmm` is
used in the backward function.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: enable autograd for torch.sparse.addmm and torch.sparse.mm"


This PR updates the derivative rule for `torch.sparse.addmm` to be
working with CSR sparse matrix. Notably `torch.sparse.sampled_addmm` is
used in the backward function.

[ghstack-poisoned]

IvanYashchuk commented:
@pytorchbot merge this please

janeyx99 commented:
@pytorchbot revert this please

as it breaks trunk 10.2 tests because sampled_addm is not available for older CUDA,
see https://hud.pytorch.org/minihud?name_filter=trunk%20/%20linux-bionic-cuda10.2-py3.9-gcc7%20/%20test%20(default,%201,%202,%20linux.4xlarge.nvidia.gpu)

IvanYashchuk committed:
Update on "Sparse CSR: enable autograd for torch.sparse.addmm and torch.sparse.mm"


This PR updates the derivative rule for `torch.sparse.addmm` to be
working with CSR sparse matrix. Notably `torch.sparse.sampled_addmm` is
used in the backward function.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: enable autograd for torch.sparse.addmm and torch.sparse.mm"


This PR updates the derivative rule for `torch.sparse.addmm` to be
working with CSR sparse matrix. Notably `torch.sparse.sampled_addmm` is
used in the backward function.

[ghstack-poisoned]

IvanYashchuk commented:
@pytorchbot merge this please

`trunk / linux-bionic-cuda10.2-py3.9-gcc7` is green now.
One job is failing and it appears to be unrelated https://github.com/pytorch/pytorch/runs/6387810616?check_suite_focus=true
